File: neetcode/121. Best Time to Buy and Sell Stock/121.java
/*
Method 1.0:
Intuition: Whenever you come across a lower value stock update buy with that price and
find the maximum profit (curr price - buy).
TC: O(n)
SC: O(1)
 */

class Solution {
    public int maxProfit(int[] prices) {
        int ans = 0;
        int buy = prices[0];

        for (int i = 0; i < prices.length; i++) {
            if (prices[i] < buy) {
                buy = prices[i];
            }

            ans = Math.max(ans, prices[i] - buy);
        }

        return ans;
    }
}

File: neetcode/121. Best Time to Buy and Sell Stock/121.kt
/*
Method 1.0: Start with a buy price and if the buy price is more than the current buy price, update
buy with that price and find the maximum profit.
TC: O(n)
SC: O(1)
 */

class Solution {
    fun maxProfit(prices: IntArray): Int {
        var ans = 0
        var buy = prices[0]

        for (i in prices.indices) {
            if (prices[i] < buy) {
                buy = prices[i]
            }

            ans = max(ans, prices[i] - buy)
        }

        return ans
    }
}
